utils/municipality.ts

```diff
@@ -0,0 +1,339 @@
+  delimiter: ";",
+  from_line: 4,
+  skip_empty_lines: true,
+  skip_lines_with_error: true,
+  trim: true
+};
+
+export interface ISerializableMunicipality {
+  codiceCatastale: string;
+  municipality: Municipality;
+}
+
+// try to decode municipality csv row in a Municipality object
+export const decodeMunicipality = (
+  record: ReadonlyArray<string>
+): t.Validation<Municipality> => {
+  if (record.length < 13) {
+    return leftE([
+      {
+        context: [],
+        value: "record has not the right length"
+      }
+    ]);
+  }
+  const municipality = {
+    codiceProvincia: record[3],
+    codiceRegione: record[0],
+    denominazione: record[5],
+    denominazioneInItaliano: record[6],
+    denominazioneRegione: record[10],
+    siglaProvincia: record[14]
+  };
+  return Municipality.decode(municipality);
+};
+
+// try to decode foreign country csv row in a Municipality object
+export const decodeForeignCountry = (
+  record: ReadonlyArray<string>
+): t.Validation<Municipality> => {
+  if (record.length < 15) {
+    return leftE([
+      {
+        context: [],
+        value: "record has not the right length"
+      }
+    ]);
+  }
+  const municipality = {
+    codiceProvincia: "",
+    codiceRegione: "",
+    denominazione: record[7],
+    denominazioneInItaliano: record[6],
+    denominazioneRegione: record[4],
+    siglaProvincia: ""
+  };
+  return Municipality.decode(municipality);
+};
+
+/**
+ * A string that represents a municipality codice catastale
+ * https://it.wikipedia.org/wiki/Codice_catastale
+ */
+
+const CODICE_CATASTALE_REGEX = `^[A-Z]\\d{3}$`;
+
+export const CodiceCatastale = PatternString(CODICE_CATASTALE_REGEX);
+export type CodiceCatastale = t.TypeOf<typeof CodiceCatastale>;
+
+/**
+ * load all the codici catastali and create a mapping between the name of the municipality and the codice catastale
+ */
+const loadMunicipalityToCatastale = (
+  blobService: BlobService
+): TaskEither<Error, Map<string, string>> =>
+  getFileFromBlob(
+    blobService,
+    MUNICIPALITIES_CONTAINER_NAME,
+    MUNICIPALITIES_CATASTALI_BLOB_NAME
+  )
+    .foldTaskEither<Error, StringMatrix>(
+      err => fromLeft(err),
+      _ =>
+        _.foldL(
+          () => fromLeft(new Error("Municipalities with catastale is empty")),
+          __ =>
+            parseCsv(__, optionMunicipalitiesWithCatastale).foldTaskEither(
+              e => fromEither(leftE(e)),
+              matrixString => fromEither(rightE(matrixString))
+            )
+        )
+    )
+    .foldTaskEither<Error, Map<string, string>>(
+      err => fromLeft(err),
+      municipalitiesCatastaleRows =>
+        fromEither(
+          rightE(
+            municipalitiesCatastaleRows.reduce(
+              (map: Map<string, string>, row) => {
+                map.set(row[1].toLowerCase(), row[0]);
+                return map;
+              },
+              new Map<string, string>()
+            )
+          )
+        )
+    );
```

Comment:
Is `foldTaskEither` needed? as you are not lifting left type, I think you can use `map`

```typescript
    .map(
      municipalitiesCatastaleRows =>
          municipalitiesCatastaleRows.reduce(
            (map: Map<string, string>, row) => {
              map.set(row[1].toLowerCase(), row[0]);
              return map;
            },
            new Map<string, string>()
          )
       );
```
(to be verified)

Reply:
I changed it into a `chain`